File: challenges/Games/Snake/snake.js

/**
 * Classic Snake Game (JavaScript Implementation)
 * ---------------------------------------------
 * Modern, well-documented, and beginner-friendly implementation of the classic Snake game for the web.
 * Features:
 * - Modular class-based design
 * - Responsive controls (WASD/Arrow keys, pause)
 * - Local storage for high score
 * - Clear comments and docstrings
 * - Optimized for readability and maintainability
 */

/**
 * Represents the Snake object, managing its position, movement, and body segments.
 */
class Snake {
    /**
     * @param {number} gridSize - The size of each grid cell in pixels.
     */
    constructor(gridSize) {
        this.gridSize = gridSize;
        this.reset();
    }

    /**
     * Resets the snake to its initial state.
     */
    reset() {
        this.x = 160;
        this.y = 160;
        this.dx = this.gridSize; // Start moving right
        this.dy = 0;
        this.cells = [];
        this.maxCells = 4;
    }

    /**
     * Updates the snake's position and body segments.
     */
    update() {
        this.x += this.dx;
        this.y += this.dy;

        // Wrap snake position horizontally on edge of screen
        if (this.x < 0) {
            this.x = this.canvasWidth - this.gridSize;
        } else if (this.x >= this.canvasWidth) {
            this.x = 0;
        }

        // Wrap snake position vertically on edge of screen
        if (this.y < 0) {
            this.y = this.canvasHeight - this.gridSize;
        } else if (this.y >= this.canvasHeight) {
            this.y = 0;
        }

        // Add new head to the beginning of the cells array
        this.cells.unshift({ x: this.x, y: this.y });

        // Remove cells as snake moves
        if (this.cells.length > this.maxCells) {
            this.cells.pop();
        }
    }

    /**
     * Draws the snake on the canvas.
     * @param {CanvasRenderingContext2D} context
     */
    draw(context) {
        context.fillStyle = 'green';
        this.cells.forEach(cell => {
            context.fillRect(cell.x, cell.y, this.gridSize - 1, this.gridSize - 1);
        });
    }

    /**
     * Handles keyboard input to change the snake's direction.
     * @param {KeyboardEvent} e
     */
    handleInput(e) {
        if ((e.key === 'a' || e.key === 'ArrowLeft') && this.dx === 0) {
            this.dx = -this.gridSize;
            this.dy = 0;
        } else if ((e.key === 'w' || e.key === 'ArrowUp') && this.dy === 0) {
            this.dy = -this.gridSize;
            this.dx = 0;
        } else if ((e.key === 'd' || e.key === 'ArrowRight') && this.dx === 0) {
            this.dx = this.gridSize;
            this.dy = 0;
        } else if ((e.key === 's' || e.key === 'ArrowDown') && this.dy === 0) {
            this.dy = this.gridSize;
            this.dx = 0;
        }
    }
}


/**
 * Represents the Apple object, managing its position.
 */
class Apple {
    /**
     * @param {number} gridSize
     * @param {number} canvasWidth
     * @param {number} canvasHeight
     */
    constructor(gridSize, canvasWidth, canvasHeight) {
        this.gridSize = gridSize;
        this.canvasWidth = canvasWidth;
        this.canvasHeight = canvasHeight;
        this.x = 0;
        this.y = 0;
        this.place();
    }

    /**
     * Places the apple at a random position on the grid.
     */
    place() {
        const maxX = this.canvasWidth / this.gridSize;
        const maxY = this.canvasHeight / this.gridSize;
        this.x = Math.floor(Math.random() * maxX) * this.gridSize;
        this.y = Math.floor(Math.random() * maxY) * this.gridSize;
    }

    /**
     * Draws the apple on the canvas.
     * @param {CanvasRenderingContext2D} context
     */
    draw(context) {
        context.fillStyle = 'red';
        context.fillRect(this.x, this.y, this.gridSize - 1, this.gridSize - 1);
    }
}


/**
 * Main Game class to orchestrate the game loop, state, and rendering.
 */
class Game {
    /**
     * @param {string} canvasId - The id of the canvas element.
     */
    constructor(canvasId) {
        this.canvas = document.getElementById(canvasId);
        this.context = this.canvas.getContext('2d');
        this.gridSize = 16;
        this.paused = false;

        this.scoreElem = document.getElementById('current-score');
        this.highScoreElem = document.getElementById('high-score');

        this.snake = new Snake(this.gridSize);
        this.snake.canvasWidth = this.canvas.width;
        this.snake.canvasHeight = this.canvas.height;

        this.apple = new Apple(this.gridSize, this.canvas.width, this.canvas.height);

        this.score = 0;
        this.highScore = Number(localStorage.getItem('snakeHighScore')) || 0;
        this.updateScoresUI();

        this.lastUpdateTime = 0;
        this.updateInterval = 100; // milliseconds between updates

        // Use arrow function to preserve 'this' context
        document.addEventListener('keydown', (e) => {
            if (e.key === 'p' || e.key === 'P') {
                this.togglePause();
            } else {
                this.snake.handleInput(e);
            }
        });
    }

    /**
     * Toggles the pause state of the game.
     */
    togglePause() {
        this.paused = !this.paused;
        if (!this.paused) {
            requestAnimationFrame((timestamp) => this.loop(timestamp));
        }
    }

    /**
     * The main game loop, called via requestAnimationFrame.
     * @param {DOMHighResTimeStamp} timestamp
     */
    loop(timestamp) {
        if (this.paused) return;

        requestAnimationFrame((ts) => this.loop(ts));

        const elapsed = timestamp - this.lastUpdateTime;
        if (elapsed > this.updateInterval) {
            this.lastUpdateTime = timestamp;
            this.update();
            this.draw();
        }
    }

    /**
     * Updates the game state (snake, collisions, etc).
     */
    update() {
        this.snake.update();
        this.checkCollisions();
    }

    /**
     * Checks for collisions (apple, self) and handles scoring and reset.
     */
    checkCollisions() {
        // Check collision with apple
        if (this.snake.cells[0].x === this.apple.x && this.snake.cells[0].y === this.apple.y) {
            this.snake.maxCells++;
            this.score++;
            this.apple.place();
            this.updateScoresUI();
        }

        // Check collision with self
        for (let i = 1; i < this.snake.cells.length; i++) {
            if (this.snake.cells[0].x === this.snake.cells[i].x && this.snake.cells[0].y === this.snake.cells[i].y) {
                this.resetGame();
            }
        }
    }

    /**
     * Resets the game state and updates high score if needed.
     */
    resetGame() {
        if (this.score > this.highScore) {
            this.highScore = this.score;
            localStorage.setItem('snakeHighScore', this.highScore);
        }
        this.score = 0;
        this.snake.reset();
        this.apple.place();
        this.updateScoresUI();
    }

    /**
     * Updates the score display in the UI.
     */
    updateScoresUI() {
        this.scoreElem.textContent = this.score;
        this.highScoreElem.textContent = this.highScore;
    }

    /**
     * Draws the game (snake and apple) on the canvas.
     */
    draw() {
        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
        this.snake.draw(this.context);
        this.apple.draw(this.context);
    }
}


// --- Main Execution ---
window.onload = () => {
    const game = new Game('game-canvas');
    requestAnimationFrame((timestamp) => game.loop(timestamp));
};

File: challenges/Games/Snake/snake.py
"""
Classic Snake Game (Turtle Graphics Implementation)
--------------------------------------------------
Modern, well-documented, and beginner-friendly implementation of the classic Snake game using Python's turtle module.
Features:
- Modular class-based design
- Clear comments and docstrings
- Optimized for readability and maintainability
"""

import time
import turtle
import random
from typing import List, Tuple

# --- Constants ---
SCREEN_WIDTH = 600
SCREEN_HEIGHT = 600
MOVE_DISTANCE = 20
STARTING_POSITIONS = [(0, 0), (-20, 0), (-40, 0)]
UP, DOWN, LEFT, RIGHT = 90, 270, 180, 0


class Snake:
    """
    Manages the snake's body, movement, and direction.
    """

    def __init__(self) -> None:
        self.segments: List[turtle.Turtle] = []
        self.create_snake()
        self.head: turtle.Turtle = self.segments[0]

    def create_snake(self) -> None:
        """
        Creates the initial snake body.
        """
        for position in STARTING_POSITIONS:
            self.add_segment(position)

    def add_segment(self, position: Tuple[int, int]) -> None:
        """
        Adds a new segment to the snake at the given position.
        """
        new_segment = turtle.Turtle("square")
        new_segment.color("white")
        new_segment.penup()
        new_segment.goto(position)
        self.segments.append(new_segment)

    def extend(self) -> None:
        """
        Adds a new segment to the end of the snake.
        """
        last_pos = self.segments[-1].position()
        self.add_segment((int(last_pos[0]), int(last_pos[1])))

    def move(self) -> None:
        """
        Moves the snake forward by one segment.
        """
        for seg_num in range(len(self.segments) - 1, 0, -1):
            new_x = self.segments[seg_num - 1].xcor()
            new_y = self.segments[seg_num - 1].ycor()
            self.segments[seg_num].goto(new_x, new_y)
        self.head.forward(MOVE_DISTANCE)

    def reset(self) -> None:
        """
        Resets the snake to its initial state.
        """
        for seg in self.segments:
            seg.goto(1000, 1000)  # Move segments off-screen
        self.segments.clear()
        self.create_snake()
        self.head = self.segments[0]

    # --- Direction Control ---
    def up(self) -> None:
        """Change direction to up if not currently moving down."""
        if self.head.heading() != DOWN:
            self.head.setheading(UP)

    def down(self) -> None:
        """Change direction to down if not currently moving up."""
        if self.head.heading() != UP:
            self.head.setheading(DOWN)

    def left(self) -> None:
        """Change direction to left if not currently moving right."""
        if self.head.heading() != RIGHT:
            self.head.setheading(LEFT)

    def right(self) -> None:
        """Change direction to right if not currently moving left."""
        if self.head.heading() != LEFT:
            self.head.setheading(RIGHT)


class Food(turtle.Turtle):
    """
    Manages the food item in the game.
    """

    def __init__(self) -> None:
        super().__init__()
        self.shape("circle")
        self.penup()
        self.shapesize(stretch_len=0.5, stretch_wid=0.5)
        self.color("red")
        self.speed("fastest")
        self.refresh()

    def refresh(self) -> None:
        """
        Moves the food to a new random location on the screen.
        """
        random_x = random.randint(-280, 280)
        random_y = random.randint(-280, 280)
        self.goto(random_x, random_y)


class Scoreboard(turtle.Turtle):
    """
    Manages the score display.
    """

    def __init__(self) -> None:
        super().__init__()
        self.score: int = 0
        self.high_score: int = 0
        self.color("white")
        self.penup()
        self.goto(0, SCREEN_HEIGHT / 2 - 40)
        self.hideturtle()
        self.update_scoreboard()

    def update_scoreboard(self) -> None:
        """
        Clears and redraws the scoreboard.
        """
        self.clear()
        self.write(
            f"Score: {self.score}  High Score: {self.high_score}",
            align="center",
            font=("Courier", 24, "normal"),
        )

    def increase_score(self) -> None:
        """
        Increments the score and updates the display.
        """
        self.score += 1
        self.update_scoreboard()

    def reset(self) -> None:
        """
        Resets the score and updates the high score if needed.
        """
        if self.score > self.high_score:
            self.high_score = self.score
        self.score = 0
        self.update_scoreboard()


class SnakeGame:
    """
    Manages the main game screen, objects, and game loop.
    """

    def __init__(self) -> None:
        self.screen = turtle.Screen()
        self.screen.setup(width=SCREEN_WIDTH, height=SCREEN_HEIGHT)
        self.screen.bgcolor("black")
        self.screen.title("Snake Game")
        self.screen.tracer(0)  # Turns off automatic screen updates

        self.snake = Snake()
        self.food = Food()
        self.scoreboard = Scoreboard()

        self.setup_key_listeners()
        self.game_is_on = True

    def setup_key_listeners(self) -> None:
        """
        Sets up the keyboard bindings for controlling the snake.
        """
        self.screen.listen()
        self.screen.onkey(self.snake.up, "w")
        self.screen.onkey(self.snake.down, "s")
        self.screen.onkey(self.snake.left, "a")
        self.screen.onkey(self.snake.right, "d")

    def run(self) -> None:
        """
        Starts and runs the main game loop.
        """
        while self.game_is_on:
            self.screen.update()
            time.sleep(0.1)
            self.snake.move()

            # Detect collision with food
            if self.snake.head.distance(self.food) < 15:
                self.food.refresh()
                self.snake.extend()
                self.scoreboard.increase_score()

            # Detect collision with wall
            if not (
                -SCREEN_WIDTH / 2 < self.snake.head.xcor() < SCREEN_WIDTH / 2
                and -SCREEN_HEIGHT / 2 < self.snake.head.ycor() < SCREEN_HEIGHT / 2
            ):
                self.scoreboard.reset()
                self.snake.reset()

            # Detect collision with tail
            for segment in self.snake.segments[1:]:
                if self.snake.head.distance(segment) < 10:
                    self.scoreboard.reset()
                    self.snake.reset()

        self.screen.exitonclick()


if __name__ == "__main__":
    game = SnakeGame()
    game.run()
